src/hooks/useVaults.tsx

```diff
@@ -0,0 +1,219 @@
+import { useApolloClient } from "@apollo/client";
+import { useEthers } from "@usedapp/core";
+import { PROTECTED_TOKENS } from "data/vaults";
+import { GET_VAULTS } from "graphql/subgraph";
+import { GET_PRICES, UniswapV3GetPrices } from "graphql/uniswap";
+import { usePrevious } from "hooks/usePrevious";
+import { useCallback, useEffect, useState, createContext, useContext } from "react";
+import { IVault, IVaultDescription } from "types/types";
+import { getTokensPrices, ipfsTransformUri } from "utils";
+
+interface IVaultsContext {
+  vaults?: IVault[]
+  tokenPrices?: number[]
+  subscribeToVaults: Function
+  removeSubscription: Function
+}
+
+export const VaultsContext = createContext<IVaultsContext>(undefined as any);
+
+export function useVaults() {
+  // every place useVaults is used we subscribed and remove subscription upon unmounting.
+  // this should not create any overhead 
+  const vaultsContext = useContext(VaultsContext);
+  useEffect(() => {
+    vaultsContext.subscribeToVaults();
+    return () => {
+      vaultsContext.removeSubscription();
+    }
+  }, [vaultsContext])
+  return vaultsContext
+}
+
+export function VaultsProvider({ children }) {
+  const [vaults, setVaults] = useState<IVault[]>();
+  const [subscriptions, setSubscrptions] = useState<number>(0)
+  const prevSubscriptions = usePrevious(subscriptions);
+  const [tokenPrices, setTokenPrices] = useState<number[]>();
+  const apolloClient = useApolloClient();
+  const { chainId } = useEthers();
+  const prevChainId = usePrevious(chainId);
+
+  // useEffect(() => {
```

Comment:
@leviyehonatan can we remove this block of comment?

Reply:
yes